src/auto-trace.js

```diff
@@ -0,0 +1,132 @@
+		callback(wrapObjectWithError(middlewareErr, stacktraceErr));
+	};	
+}
+
+/**
+ * Wraps rawError in an Error object (if typeOf rawError =! Error) using the async stacktrace
+ * Calls globalMiddleware and extraMiddleware functions on rawError before wrapping in Error object 
+ * Behaves exactly as asyncStacktrace, but throws the error object instead of returning.
+ * In most cases the async stacktrace is most useful - it contains the stacktrace before switching context to caller timeout 
+ *
+ * @param {Object}   [rawError] 
+ * @param {function (Object err) => Object newErr } [extraMiddlewares] - middleware function(s) that will be called on rawError before being handled
+ * @throws {Error} 
+ */
+export function throwAsyncStacktrace(...extraMiddlewares) {
+	const stacktraceErr = new Error();
+
+	return (rawError) => {
+		const err = rawError ? rawError : new Error();
+		const middlewareErr = executeMiddleware(err, extraMiddlewares);
+		throw wrapObjectWithError(middlewareErr, stacktraceErr);
+	};
+}
+
+/**
+ * Wraps rawError in an Error object (if typeOf rawError =! Error) using the defualt stacktrace
+ * Calls globalMiddleware and extraMiddleware functions on rawError before wrapping in Error object
+ * sync stacktrace contains the stacktrace afer switching context to caller timeout 
+ *
+ * @param {Object}   [rawError] 
+ * @param {function (Object err) => Object newErr } [extraMiddlewares] - middleware function(s) that will be called on rawError before being handled
+ * @returns {Error} 
+ */
+export function syncStacktrace(rawError, ...extraMiddlewares){
+	const err = rawError ? rawError : new Error();
+	const middlewareErr = executeMiddleware(err, extraMiddlewares);
+	return wrapObjectWithError(middlewareErr);
+}
+
+/**
+ * Wraps rawError in an Error object (if typeOf rawError =! Error) using the defualt stacktrace
+ * Calls globalMiddleware and extraMiddleware functions on rawError before wrapping in Error object 
+ * Behaves exactly as syncStacktrace, but throws the error object instead of returning.
+ * sync stacktrace contains the stacktrace afer switching context to caller timeout 
+ * 
+ * @param {Object}   [rawError] 
+ * @param {function (Object err) => Object newErr } [extraMiddlewares] - middleware function(s) that will be called on rawError before being handled
+ * @throws {Error} 
+ */
+export function throwSyncStacktrace(rawError, ...extraMiddlewares){
+	throw syncStacktrace(rawError, ...extraMiddlewares);
+}
+
+/**
+ * Calls globalMiddleware and extraMiddleware functions on rawError
+ * Does not create a stacktrace nor will it wrap rawError in Error object
+ *
+ * @param {Object}   [rawError] 
+ * @param {function (Object err) => Object newErr } [extraMiddlewares] - middleware function(s) that will be called on rawError
+ * @returns {Object} 
+ */
+export function withoutStacktrace(rawError, ...extraMiddlewares){
+	const err = rawError ? rawError : new Error();
```

Comment:
I feel like we talked about this yesterday and I just forgot, but do we want to default to an empty error object? Usually the use case for this function is when you want to not create an Error object, right?